modules/nlu/src/backend/pipelines/entities/pattern_extractor.ts

```diff
@@ -1,71 +1,135 @@
-          end: res.sourceIndex + res.value.length,
-          raw: {}
-        },
-        data: {
-          extras: {},
-          value: res.value,
-          unit: 'string'
+
+import { flatMap, flatten } from 'lodash'
+
+import { extractPattern } from '../../tools/patterns-utils'
+import { tokenize } from '../language/tokenizers'
+
+const debug = DEBUG('nlu').sub('entities')
+const debugLists = debug.sub('lists')
+
+export default class PatternExtractor {
+  constructor(private toolkit: typeof sdk.MLToolkit) {}
+
+  async extractLists(input: string, lang: string, entityDefs: sdk.NLU.EntityDefinition[]): Promise<sdk.NLU.Entity[]> {
+    return flatten(
+      flatten(
+        await Promise.map(entityDefs, async entityDef => {
+          return await Promise.map(entityDef.occurences || [], occurence =>
+            this._extractEntitiesFromOccurence(input, lang, occurence, entityDef)
+          )
+        })
+      )
+    )
+  }
+
+  protected async _extractEntitiesFromOccurence(
+    input: string,
+    lang: string,
+    occurence: sdk.NLU.EntityDefOccurence,
+    entityDef: sdk.NLU.EntityDefinition
+  ): Promise<sdk.NLU.Entity[]> {
+    const tokens = await tokenize(input, lang)
+    const values = await Promise.map([occurence.name, ...occurence.synonyms], v => tokenize(v, lang))
+
+    const findings: sdk.NLU.Entity[] = []
+
+    let cur = 0
+    for (const tok of tokens) {
+      cur = cur + input.substr(cur).indexOf(tok)
+
+      let highest = 0
+      let extracted = ''
+      let source = ''
+
+      for (const val of values) {
+        let partOfPhrase = tok
+        const occ = val.join('+')
+        if (val.length > 1) {
+          partOfPhrase = (await tokenize(input.substr(cur, occ.length), lang)).join('+')
         }
-      }))
-    } catch (error) {
-      throw Error(`Pattern of entity ${entityDef.name} is invalid`)
-    }
-  })
-}
 
-const _extractEntitiesFromOccurence = (
-  input: string,
-  occurence: sdk.NLU.EntityDefOccurence,
-  entityDef: sdk.NLU.EntityDefinition
-): sdk.NLU.Entity[] => {
-  const pattern = [occurence.name, ...occurence.synonyms].map(escapeRegex).join('|')
-
-  try {
-    const regex = new RegExp(pattern, 'i')
-    return extractPattern(input, regex).map(extracted => ({
-      name: entityDef.name,
-      type: 'list',
-      meta: {
-        confidence: 1, // extrated with synonyme as patterns
-        provider: 'native',
-        source: extracted.value,
-        start: extracted.sourceIndex,
-        end: extracted.sourceIndex + extracted.value.length,
-        raw: {}
-      },
-      data: {
-        extras: {},
-        value: occurence.name, // cannonical value,
-        unit: 'string'
+        let distance = 0.0
+
+        if (entityDef.fuzzy) {
+          const d1 = this.toolkit.Strings.computeLevenshteinDistance(partOfPhrase, occ)
+          const d2 = this.toolkit.Strings.computeJaroWinklerDistance(partOfPhrase, occ, { caseSensitive: true })
+          distance = Math.min(d1, d2)
```

Comment:
Wouldn't it better if you were pessimist on this and use the max ? you could also use an avg so no optimism, no pessimism only "maths"

Reply:
Tried both and honestly I don't think that's a good idea

Reply:
I agree that avg is probably not a good idea. But how taking the min dst is being pessimistic ?